### ISSUE TITLE
misc-redundant-expression: expression with calls to a non-pure function is considered as always-true

### ISSUE THREAD

**Issue by pavelkryukov** [llvm/llvm-project#35853]
|  |  |
| --- | --- |
| Bugzilla Link | [36505](https://llvm.org/bz36505) |
| Version | unspecified |
| OS | Windows NT |
| Attachments | [Example of always-true false positive with calls to a non-pure function](https://user-images.githubusercontent.com/92601623/143757131-cbdb4144-d0a7-4c13-84d6-c4843f4b39ac.gz) |
| CC | @EugeneZelenko,@Xazax-hun |

## Extended Description 

Clang-Tidy produces a false positive warning about always-true expression (misc-redundant-expression check) if expression contains consequent calls to a non-pure function which have different return values:

```
$> clang-tidy -checks='*' ldfile.c -- -Wall
15 warnings generated.
./ldfile.c:11:32: warning: logical expression is always true [misc-redundant-expression]
    if ((token = yylex()) != 1 || (token = yylex()) != 2) {
                               ^
Suppressed 14 warnings (14 in non-user code).

$> clang ldfile.c -Wall -Wextra && ./a.out
false
```

The code is taken from GNU Binutils (/ld/ldfile.c) source file.

**Comment by pavelkryukov** [comment]
@EugeneZelenko @Xazax-hun I’ll try to get more attention, as wrong fix-it can be generated because of this.

**Comment by FabianWolff** [comment]
This is unrelated to #54011 and even intended to some extent, I believe: Almost every function call might have side effects (and `misc-redundant-expression` does not look into the called function), but taking this into account would prevent many legitimate warnings (arising from typos etc.) from firing, so the false positives here are tolerated.

Instead, to fix your example, I'd focus on the assignments. I've opened [D122535](https://reviews.llvm.org/D122535) to this end.

**Comment by pavelkryukov** [comment]
Thank you!
